zjit/src/codegen.rs

```diff
@@ -220,14 +220,26 @@ fn gen_iseq(cb: &mut CodeBlock, iseq: IseqPtr, function: Option<&Function>) -> R
 
 /// Compile an ISEQ into machine code
 fn gen_iseq_body(cb: &mut CodeBlock, iseq: IseqPtr, function: Option<&Function>, payload: &mut IseqPayload) -> Result<IseqCodePtrs, CompileError> {
+    // If we ran out of code region, we shouldn't attempt to generate new code.
+    if cb.has_dropped_bytes() {
+        return Err(CompileError::OutOfMemory);
+    }
+
     // Convert ISEQ into optimized High-level IR if not given
     let function = match function {
         Some(function) => function,
         None => &compile_iseq(iseq)?,
     };
 
     // Compile the High-level IR
-    let (iseq_code_ptrs, gc_offsets, iseq_calls) = gen_function(cb, iseq, function)?;
+    let (iseq_code_ptrs, gc_offsets, iseq_calls) = gen_function(cb, iseq, function).inspect_err(|err| {
+        // If we use too much memory to compile this ISEQ, it would set cb.dropped_bytes = true.
+        // To avoid failing future compilation due to cb.has_dropped_bytes(), check the threshold
+        // again with the current zjit_alloc_bytes() which may be decreased after gen_function().
+        if *err == CompileError::OutOfMemory {
```

Comment:
This feels like it should be part of the assembler or codeblock API; it should leave the assembler in a good state:

1. generated function => return ok
2. generated function, but no more bytes left => return ok
3. partially generated function, no more bytes left => reset to whatever it was before starting this function, return err
4. no more bytes left => return err

Reply:
Perhaps as soon as we OOM we can stomp on the entry trampoline

Reply:
> But as of asm.compile(), the Assembler::insns is still alive, so it's not possible.

I just realized this assumption wasn't true. Once `asm.compile()` is done with `self.compile_with_regs()`, the `Assembler` is already moved to `compile_with_regs` and dropped. So we can do it inside `asm.compile()`. I'll fix it that way.